subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.hash.HashCode;
+import org.gradle.api.Task;
+import org.gradle.api.internal.tasks.execution.TaskCachingHashesListener;
+import org.gradle.caching.BuildCacheKey;
+import org.gradle.caching.internal.BuildCacheKeyBuilder;
+import org.gradle.caching.internal.DefaultBuildCacheKeyBuilder;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class TaskCacheKeyCalculator {
+    private final TaskCachingHashesListener hashesListener;
+
+    public TaskCacheKeyCalculator() {
+        this(null);
+    }
+
+    public TaskCacheKeyCalculator(TaskCachingHashesListener hashesListener) {
+        this.hashesListener = hashesListener;
+    }
+
+    public BuildCacheKey calculate(TaskExecution execution, Task task) {
```

Comment:
I don't like that it's mutable, but I was thinking about `LazyTaskExecution` more than just `TaskExecution`.  It's all really hard to follow when a particular value is going to be populated or not.  

I don't feel very strongly about this.  I also didn't want `TaskExecution` being used too widely (again, since it's hard to tell when something might be added/modified).